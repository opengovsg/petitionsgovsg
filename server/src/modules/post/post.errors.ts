import { StatusCodes } from 'http-status-codes'
import { ApplicationError } from '../core/core.errors'

export class MissingPublicPostError extends ApplicationError {
  constructor(
    message = 'No public post with this id',
    statusCode = StatusCodes.NOT_FOUND,
  ) {
    super(message, statusCode)
  }
}

export class PostUpdateError extends ApplicationError {
  constructor(
    message = 'Post update failed',
    statusCode = StatusCodes.INTERNAL_SERVER_ERROR,
  ) {
    super(message, statusCode)
  }
}
