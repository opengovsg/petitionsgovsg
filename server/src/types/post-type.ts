import { Post } from '~shared/types/base'

export type PostEditType = {
  id: number
  userid: number
  summary: string
  title: string
}

export type UpdatePostRequestDto = Pick<Post, 'title' | 'summary'>
