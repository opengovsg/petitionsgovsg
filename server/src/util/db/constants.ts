export const POST_ID = 'af7b6aa9-7c6e-4ea5-b536-4e55bc506847'
