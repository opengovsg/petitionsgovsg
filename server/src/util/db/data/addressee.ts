export const mockAddressee = {
  name: 'Ministry of Environment',
  shortName: 'MoE',
}

export const mockAddressees = [
  {
    id: 1,
    name: 'Ministry of Environment',
    shortName: 'MoE',
    createdAt: '2022-01-01T00:00:00+00:00',
    updatedAt: '2022-01-01T00:00:00+00:00',
    openToPublic: false,
    officerEmails: [],
  },
  {
    id: 2,
    name: 'Ministry of Labour',
    shortName: 'MoL',
    createdAt: '2022-01-01T00:00:00+00:00',
    updatedAt: '2022-01-01T00:00:00+00:00',
    openToPublic: false,
    officerEmails: [],
  },
]
