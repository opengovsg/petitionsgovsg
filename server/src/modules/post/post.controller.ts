import { validationResult } from 'express-validator'
import { StatusCodes } from 'http-status-codes'
import { Post } from '~shared/types/base'
import { createLogger } from '../../bootstrap/logging'
import { Message } from '../../types/message-type'
import { ControllerHandler } from '../../types/response-handler'
import { SortType } from '../../types/sort-type'
import { AuthService } from '../auth/auth.service'
import { PostService, PostWithAddresseeAndSignatures } from './post.service'
import { hashData, generateSalt } from '../../util/hash'

const logger = createLogger(module)

export class PostController {
  private authService: Public<AuthService>
  private postService: Public<PostService>

  constructor({
    authService,
    postService,
  }: {
    authService: Public<AuthService>
    postService: Public<PostService>
  }) {
    this.authService = authService
    this.postService = postService
  }

  /**
   * Lists all post
   * @query sort Sort by popularity or recent
   * @query size Number of posts to return
   * @query page If size is given, specify which page to return
   * @return 200 with posts and totalItem for pagination
   * @return 500 when database error occurs
   */
  listPosts: ControllerHandler<
    undefined,
    { posts: Post[]; totalItems: number } | Message,
    undefined,
    {
      page?: number
      size?: number
      sort?: SortType
    }
  > = async (req, res) => {
    const { page, size, sort = SortType.Top } = req.query
    try {
      const data = await this.postService.listPosts({
        sort: sort as SortType,
        page: page,
        size: size,
      })
      return res.status(StatusCodes.OK).json(data)
    } catch (error) {
      if (error instanceof Error) {
        logger.error({
          message: 'Error while listing posts',
          meta: {
            function: 'listPosts',
          },
          error,
        })
        return res
          .status(StatusCodes.INTERNAL_SERVER_ERROR)
          .json({ message: 'Server Error' })
      }
    }
  }

  /**
   * Get a single post and all the users associated with it
   * @param postId Id of the post
   * @query relatedPosts if true, return related posts
   * @return 200 with post
   * @return 403 if user does not have permission to access post
   * @return 500 for database error
   */
  getSinglePost: ControllerHandler<
    { id: number },
    PostWithAddresseeAndSignatures | Message,
    undefined,
    { relatedPosts?: number }
  > = async (req, res) => {
    const errors = validationResult(req)
    if (!errors.isEmpty()) {
      return res.status(StatusCodes.BAD_REQUEST).json(errors.array()[0].msg)
    }
    let post
    try {
      post = await this.postService.getSinglePost(req.params.id)
    } catch (error) {
      logger.error({
        message: 'Error while retrieving single post',
        meta: {
          function: 'getSinglePost',
        },
        error,
      })
      return res
        .status(StatusCodes.INTERNAL_SERVER_ERROR)
        .json({ message: 'Server Error' })
    }

    try {
      await this.authService.verifyUserCanViewPost(
        post,
        req.user?.id.toString() ?? '',
      )
    } catch (error) {
      logger.error({
        message: 'Error while retrieving single post',
        meta: {
          function: 'getSinglePost',
        },
        error,
      })
      return res
        .status(StatusCodes.FORBIDDEN)
        .json({ message: 'User does not have permission to access this post' })
    }

    return res.status(StatusCodes.OK).json(post)
  }

  /**
   * Create a new post
   * @body title title of post
   * @body summary summary of post
   * @body reason reason of post
   * @body request request of post
   * @body references references of post
   * @return 200 if post is created
   * @return 400 if title and description is too short or long
   * @return 401 if user is not signed in
   * @return 500 if database error
   */
  createPost: ControllerHandler<
    undefined,
    { data: number } | Message,
    {
      title: string
      summary: string | null
      reason: string
      request: string
      references: string | null
      fullname: string
      addresseeId: number
      profile: string | null
      email: string
      salt: string
    },
    undefined
  > = async (req, res) => {
    const errors = validationResult(req)
    if (!errors.isEmpty()) {
      return res.status(StatusCodes.BAD_REQUEST).json(errors.array()[0].msg)
    }
    if (!req.user) {
      return res
        .status(StatusCodes.UNAUTHORIZED)
        .json({ message: 'User not signed in' })
    }

    try {
      if (!req.user) {
        return res
          .status(StatusCodes.UNAUTHORIZED)
          .json({ message: 'User not signed in' })
      }

      const salt = await generateSalt()
      const hashedUserSgid = await hashData(req.user.id, salt)
      const data = await this.postService.createPost({
        title: req.body.title,
        summary: req.body.summary,
        reason: req.body.reason,
        request: req.body.request,
        hashedUserSgid: hashedUserSgid,
        references: req.body.references,
        fullname: req.body.fullname,
        addresseeId: req.body.addresseeId,
        profile: req.body.profile,
        email: req.body.email,
        salt: salt,
      })

      return res.status(StatusCodes.OK).json({ data: data })
    } catch (error) {
      logger.error({
        message: 'Error while creating post',
        meta: {
          function: 'addPost',
        },
        error,
      })
      return res
        .status(StatusCodes.INTERNAL_SERVER_ERROR)
        .json({ message: 'Server error' })
    }
  }

  /**
   * Delete a post
   * @param id Post to be deleted
   * @return 200 if successful
   * @return 401 if user is not logged in
   * @return 403 if user does not have permission to delete post
   * @return 500 if database error
   */
  deletePost: ControllerHandler<{ id: string }, Message> = async (req, res) => {
    const postId = Number(req.params.id)
    try {
      const userId = req.user?.id
      if (!userId) {
        logger.error({
          message: 'UserId is undefined after authenticated',
          meta: {
            function: 'deletePost',
          },
        })
        return res
          .status(StatusCodes.UNAUTHORIZED)
          .json({ message: 'You must be logged in to delete posts.' })
      }
      // TODO
      // const hasPermission = await this.authService.hasPermissionToEditPost(
      //   userId,
      //   postId,
      // )
      // if (!hasPermission) {
      //   return res
      //     .status(StatusCodes.FORBIDDEN)
      //     .json({ message: 'You do not have permission to delete this post.' })
      // }
      await this.postService.deletePost(postId)
      return res.status(StatusCodes.OK).send({ message: 'OK' })
    } catch (error) {
      logger.error({
        message: 'Error while deleting post',
        meta: {
          function: 'deletePost',
        },
        error,
      })
      return res
        .status(StatusCodes.INTERNAL_SERVER_ERROR)
        .json({ message: 'Server error' })
    }
  }
}
