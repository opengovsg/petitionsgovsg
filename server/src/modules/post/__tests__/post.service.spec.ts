import {
  createTestDatabase,
  resetAndSetupDb,
  SequelizeWithModels,
} from '@/util/db/jest-db'
import { PostService } from '../post.service'
import { SortType } from '~shared/types/base'
import { mockPost } from '@/util/db/data/post'
import { INVALID_POST_ID, POST_ID } from '@/util/db/constants'
import { Post, PostStatus } from '~shared/types/base'
import {
  MissingPublicPostError,
  PostUpdateError,
  AddresseeDoesNotExistError,
} from '../post.errors'

/**
 * TODO: getPaginatedPosts
 */

describe('PostService', () => {
  let service: PostService
  let db: SequelizeWithModels

  beforeAll(async () => {
    db = await createTestDatabase()
    service = new PostService({
      Signature: db.Signature,
      Post: db.Post,
      Addressee: db.Addressee,
      sequelize: db.sequelize,
    })
  })

  beforeEach(async () => {
    await resetAndSetupDb(db)
    jest.resetAllMocks()
  })

  describe('listPosts', () => {
    beforeEach(async () => {
      const newlyCreatedPost = {
        ...mockPost,
        createdAt: '2022-02-02T00:00:00.000Z',
        updatedAt: '2022-02-02T00:00:00.000Z',
      }
      await db.Post.create(newlyCreatedPost, { silent: true })
    })

    it('returns posts ordered by date created', async () => {
      const post = await service.listPosts({ sort: SortType.Newest })

      expect(post).toMatchSnapshot()
    })

    it('returns posts ordered by date created (oldest)', async () => {
      const post = await service.listPosts({ sort: SortType.Oldest })

      expect(post).toMatchSnapshot()
    })

    it('returns posts ordered by signature count', async () => {
      const post = await service.listPosts({ sort: SortType.MostSignatures })

      expect(post).toMatchSnapshot()
    })

    it('returns posts ordered by signature count (least)', async () => {
      const post = await service.listPosts({ sort: SortType.LeastSignatures })

      expect(post).toMatchSnapshot()
    })
  })

  describe('getSinglePost', () => {
    let closedPost: Post
    beforeEach(async () => {
      const mockClosedPost = {
        ...mockPost,
        status: PostStatus.Closed,
      }
      closedPost = await db.Post.create(mockClosedPost)
    })
    it('returns a single post with specified post id', async () => {
      const post = await service.getSinglePost(POST_ID)

      expect(post).toMatchSnapshot()
    })

    it('throws an error if post is not public', async () => {
      const postError = service.getSinglePost(closedPost.id)

      await expect(postError).rejects.toStrictEqual(
        new MissingPublicPostError(),
      )
      await expect(postError).rejects.toThrowErrorMatchingSnapshot()
    })
  })

  describe('createPost', () => {
    it('creates post on good input', async () => {
      const postParams = {
        title: 'Title should be concise',
        summary: 'A short description or summary ',
        reason: 'Reason for post goes here',
        request: 'Request for post goes her',
        hashedUserSgid: 'wxyz',
        fullname: 'John Doe',
        addresseeId: 1,
        email: 'john@email.com',
        salt: 'abcdef',
        profile: null,
        references: null,
      }

      const postId = await service.createPost(postParams)

      const post = await db.Post.findByPk(postId)
      expect(post).toBeDefined()
      expect(post?.get()).toMatchSnapshot({
        id: expect.any(String),
        createdAt: expect.any(Date),
        updatedAt: expect.any(Date),
      })
    })

    it('throws on bad addressee', async () => {
      const postParams = {
        title: 'Title should be concise',
        summary: 'A short description or summary',
        reason: 'Reason for post goes here',
        request: 'Request for post goes her',
        hashedUserSgid: 'wxyz',
        fullname: 'John Doe',
        addresseeId: -1,
        email: 'john@email.com',
        salt: 'abcdef',
        profile: null,
        references: null,
      }

      const postError = service.createPost(postParams)

      await expect(postError).rejects.toStrictEqual(
        new AddresseeDoesNotExistError(),
      )
      await expect(postError).rejects.toThrowErrorMatchingSnapshot()
    })
  })

  describe('updatePost', () => {
    it('updates post on good input', async () => {
      const postParams = {
        id: POST_ID,
        title: 'Title should be even more concise',
        summary: 'A longer description or summary',
        reason: 'More compelling reason',
        request: 'More compelling request',
        references: null,
        addresseeId: 2,
        profile: 'User profile',
        email: 'updated@email.com',
      }

      const postUpdateStatus = await service.updatePost(postParams)

      const updatedPost = await db.Post.findByPk(POST_ID)

      expect(postUpdateStatus).toBeTruthy()
      expect(updatedPost).toMatchObject(postParams)
      expect(updatedPost?.get()).toMatchSnapshot({
        createdAt: expect.any(Date),
        updatedAt: expect.any(Date),
      })
    })

    it('throws on bad addressee', async () => {
      const postParams = {
        id: POST_ID,
        title: 'Title should be even more concise',
        summary: 'A longer description or summary',
        reason: 'More compelling reason',
        request: 'More compelling request',
        references: null,
        addresseeId: -1,
        profile: 'User profile',
        email: 'updated@email.com',
      }

      const updateError = service.updatePost(postParams)

      await expect(updateError).rejects.toStrictEqual(
        new AddresseeDoesNotExistError(),
      )
      await expect(updateError).rejects.toThrowErrorMatchingSnapshot()
    })
  })

  describe('deletePost', () => {
    it('archives post successfully', async () => {
      await service.deletePost(POST_ID)
      const archivedPost = await db.Post.findByPk(POST_ID)

      expect(archivedPost?.status).toEqual(PostStatus.Closed)
      expect(archivedPost?.get()).toMatchSnapshot({
        createdAt: expect.any(Date),
        updatedAt: expect.any(Date),
      })
    })

    it('throws an error on invalid id', async () => {
      const archiveError = service.deletePost(INVALID_POST_ID)

      await expect(archiveError).rejects.toStrictEqual(new PostUpdateError())
      await expect(archiveError).rejects.toThrowErrorMatchingSnapshot()
    })
  })

  describe('publishPost', () => {
    let closedPost: Post
    beforeEach(async () => {
      const mockClosedPost = {
        ...mockPost,
        status: PostStatus.Closed,
      }
      closedPost = await db.Post.create(mockClosedPost)
    })
    it('returns void on making post public', async () => {
      await service.publishPost(closedPost.id)
      const publishedPost = await db.Post.findByPk(closedPost.id)

      expect(publishedPost?.status).toEqual(PostStatus.Open)
      expect(publishedPost?.get()).toMatchSnapshot({
        createdAt: expect.any(Date),
        updatedAt: expect.any(Date),
      })
    })

    it('throws an error on invalid id', async () => {
      const publishError = service.publishPost(INVALID_POST_ID)

      await expect(publishError).rejects.toStrictEqual(new PostUpdateError())
      await expect(publishError).rejects.toThrowErrorMatchingSnapshot()
    })
  })
})
