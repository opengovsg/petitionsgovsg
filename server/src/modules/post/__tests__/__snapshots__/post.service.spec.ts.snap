// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostService createPost creates post on good input 1`] = `
Object {
  "addresseeId": 1,
  "auditTrail": Object {},
  "createdAt": Any<Date>,
  "email": "john@email.com",
  "fullname": "John Doe",
  "hashedUserSgid": "wxyz",
  "id": Any<String>,
  "profile": null,
  "reason": "Reason for post goes here",
  "references": null,
  "request": "Request for post goes her",
  "salt": "abcdef",
  "signatureOptions": Array [
    "support",
    "oppose",
  ],
  "status": "DRAFT",
  "summary": "A short description or summary ",
  "title": "Title should be concise",
  "updatedAt": Any<Date>,
}
`;

exports[`PostService createPost throws on bad addressee 1`] = `"Addressee does not exist"`;

exports[`PostService deletePost archives post successfully 1`] = `
Object {
  "addresseeId": 1,
  "auditTrail": Object {},
  "createdAt": Any<Date>,
  "email": "John@email.com",
  "fullname": "John Doe",
  "hashedUserSgid": "$2b$10$72TKINiR4.ZMaJZmuN5tW.uCuGZcA0TzLUo4qly55GyDndAYG8BSW",
  "id": "af7b6aa9-7c6e-4ea5-b536-4e55bc506847",
  "profile": null,
  "reason": "reason for",
  "references": null,
  "request": "request",
  "salt": "$2b$10$72TKINiR4.ZMaJZmuN5tW.",
  "signatureOptions": Array [
    "support",
    "oppose",
  ],
  "status": "CLOSED",
  "summary": null,
  "title": "Basic petition without optional items",
  "updatedAt": Any<Date>,
}
`;

exports[`PostService deletePost throws an error on invalid id 1`] = `"Post update failed"`;

exports[`PostService getSinglePost returns a single post with specified post id 1`] = `
Object {
  "addressee": Object {
    "name": "Ministry of Environment",
    "shortName": "MoE",
  },
  "addresseeId": 1,
  "createdAt": 2022-01-01T00:00:00.000Z,
  "email": "John@email.com",
  "fullname": "John Doe",
  "hashedUserSgid": "$2b$10$72TKINiR4.ZMaJZmuN5tW.uCuGZcA0TzLUo4qly55GyDndAYG8BSW",
  "profile": null,
  "reason": "reason for",
  "references": null,
  "request": "request",
  "salt": "$2b$10$72TKINiR4.ZMaJZmuN5tW.",
  "signatureCount": 3,
  "signatures": Array [
    Object {
      "comment": null,
      "createdAt": 2022-01-01T00:00:00.000Z,
      "fullname": null,
      "hashedUserSgid": "1234",
      "id": 1,
      "postId": "af7b6aa9-7c6e-4ea5-b536-4e55bc506847",
      "type": "support",
      "updatedAt": 2022-01-01T00:00:00.000Z,
    },
    Object {
      "comment": null,
      "createdAt": 2022-01-01T00:00:00.000Z,
      "fullname": "John Doe",
      "hashedUserSgid": "5678",
      "id": 2,
      "postId": "af7b6aa9-7c6e-4ea5-b536-4e55bc506847",
      "type": "support",
      "updatedAt": 2022-01-01T00:00:00.000Z,
    },
    Object {
      "comment": "I support this cause!",
      "createdAt": 2022-01-01T00:00:00.000Z,
      "fullname": null,
      "hashedUserSgid": "8765",
      "id": 3,
      "postId": "af7b6aa9-7c6e-4ea5-b536-4e55bc506847",
      "type": "support",
      "updatedAt": 2022-01-01T00:00:00.000Z,
    },
  ],
  "status": "OPEN",
  "summary": null,
  "title": "Basic petition without optional items",
  "updatedAt": 2022-01-01T00:00:00.000Z,
}
`;

exports[`PostService getSinglePost throws an error if post is not public 1`] = `"No public post with this id"`;

exports[`PostService listPosts returns posts ordered by date created (oldest) 1`] = `
Object {
  "posts": Array [
    Object {
      "addressee": Object {
        "name": "Ministry of Environment",
        "shortName": "MoE",
      },
      "addresseeId": 1,
      "createdAt": 2022-01-01T00:00:00.000Z,
      "email": "John@email.com",
      "fullname": "John Doe",
      "hashedUserSgid": "$2b$10$72TKINiR4.ZMaJZmuN5tW.uCuGZcA0TzLUo4qly55GyDndAYG8BSW",
      "id": "af7b6aa9-7c6e-4ea5-b536-4e55bc506847",
      "profile": null,
      "reason": "reason for",
      "references": null,
      "request": "request",
      "signatureCount": 3,
      "signatures": Array [
        Object {
          "comment": null,
          "createdAt": 2022-01-01T00:00:00.000Z,
          "fullname": null,
          "hashedUserSgid": "1234",
          "id": 1,
          "postId": "af7b6aa9-7c6e-4ea5-b536-4e55bc506847",
          "type": "support",
          "updatedAt": 2022-01-01T00:00:00.000Z,
        },
        Object {
          "comment": null,
          "createdAt": 2022-01-01T00:00:00.000Z,
          "fullname": "John Doe",
          "hashedUserSgid": "5678",
          "id": 2,
          "postId": "af7b6aa9-7c6e-4ea5-b536-4e55bc506847",
          "type": "support",
          "updatedAt": 2022-01-01T00:00:00.000Z,
        },
        Object {
          "comment": "I support this cause!",
          "createdAt": 2022-01-01T00:00:00.000Z,
          "fullname": null,
          "hashedUserSgid": "8765",
          "id": 3,
          "postId": "af7b6aa9-7c6e-4ea5-b536-4e55bc506847",
          "type": "support",
          "updatedAt": 2022-01-01T00:00:00.000Z,
        },
      ],
      "status": "OPEN",
      "summary": null,
      "title": "Basic petition without optional items",
      "updatedAt": 2022-01-01T00:00:00.000Z,
    },
    Object {
      "addressee": Object {
        "name": "Ministry of Environment",
        "shortName": "MoE",
      },
      "addresseeId": 1,
      "createdAt": 2022-01-01T00:00:00.000Z,
      "email": "John@email.com",
      "fullname": "John Doe",
      "hashedUserSgid": "wxyz",
      "id": "aaaaaaaa-dea2-4f78-8dc4-5542cafaff87",
      "profile": "OGP",
      "reason": "reason",
      "references": "wikipedia.com",
      "request": "request",
      "signatureCount": 0,
      "signatures": Array [],
      "status": "OPEN",
      "summary": "petition summary",
      "title": "Petition with summary, references, profile",
      "updatedAt": 2022-01-01T00:00:00.000Z,
    },
    Object {
      "addressee": Object {
        "name": "Ministry of Environment",
        "shortName": "MoE",
      },
      "addresseeId": 1,
      "createdAt": 2022-02-02T00:00:00.000Z,
      "email": "John@email.com",
      "fullname": "John Doe",
      "hashedUserSgid": "wxyz",
      "id": "395ac6ec-dea2-4f78-8dc4-1234cafaff87",
      "profile": null,
      "reason": "reason for",
      "references": null,
      "request": "request",
      "signatureCount": 0,
      "signatures": Array [],
      "status": "OPEN",
      "summary": null,
      "title": "Basic petition without optional items",
      "updatedAt": 2022-02-02T00:00:00.000Z,
    },
  ],
  "totalItems": 3,
}
`;

exports[`PostService listPosts returns posts ordered by date created 1`] = `
Object {
  "posts": Array [
    Object {
      "addressee": Object {
        "name": "Ministry of Environment",
        "shortName": "MoE",
      },
      "addresseeId": 1,
      "createdAt": 2022-02-02T00:00:00.000Z,
      "email": "John@email.com",
      "fullname": "John Doe",
      "hashedUserSgid": "wxyz",
      "id": "395ac6ec-dea2-4f78-8dc4-1234cafaff87",
      "profile": null,
      "reason": "reason for",
      "references": null,
      "request": "request",
      "signatureCount": 0,
      "signatures": Array [],
      "status": "OPEN",
      "summary": null,
      "title": "Basic petition without optional items",
      "updatedAt": 2022-02-02T00:00:00.000Z,
    },
    Object {
      "addressee": Object {
        "name": "Ministry of Environment",
        "shortName": "MoE",
      },
      "addresseeId": 1,
      "createdAt": 2022-01-01T00:00:00.000Z,
      "email": "John@email.com",
      "fullname": "John Doe",
      "hashedUserSgid": "$2b$10$72TKINiR4.ZMaJZmuN5tW.uCuGZcA0TzLUo4qly55GyDndAYG8BSW",
      "id": "af7b6aa9-7c6e-4ea5-b536-4e55bc506847",
      "profile": null,
      "reason": "reason for",
      "references": null,
      "request": "request",
      "signatureCount": 3,
      "signatures": Array [
        Object {
          "comment": null,
          "createdAt": 2022-01-01T00:00:00.000Z,
          "fullname": null,
          "hashedUserSgid": "1234",
          "id": 1,
          "postId": "af7b6aa9-7c6e-4ea5-b536-4e55bc506847",
          "type": "support",
          "updatedAt": 2022-01-01T00:00:00.000Z,
        },
        Object {
          "comment": null,
          "createdAt": 2022-01-01T00:00:00.000Z,
          "fullname": "John Doe",
          "hashedUserSgid": "5678",
          "id": 2,
          "postId": "af7b6aa9-7c6e-4ea5-b536-4e55bc506847",
          "type": "support",
          "updatedAt": 2022-01-01T00:00:00.000Z,
        },
        Object {
          "comment": "I support this cause!",
          "createdAt": 2022-01-01T00:00:00.000Z,
          "fullname": null,
          "hashedUserSgid": "8765",
          "id": 3,
          "postId": "af7b6aa9-7c6e-4ea5-b536-4e55bc506847",
          "type": "support",
          "updatedAt": 2022-01-01T00:00:00.000Z,
        },
      ],
      "status": "OPEN",
      "summary": null,
      "title": "Basic petition without optional items",
      "updatedAt": 2022-01-01T00:00:00.000Z,
    },
    Object {
      "addressee": Object {
        "name": "Ministry of Environment",
        "shortName": "MoE",
      },
      "addresseeId": 1,
      "createdAt": 2022-01-01T00:00:00.000Z,
      "email": "John@email.com",
      "fullname": "John Doe",
      "hashedUserSgid": "wxyz",
      "id": "aaaaaaaa-dea2-4f78-8dc4-5542cafaff87",
      "profile": "OGP",
      "reason": "reason",
      "references": "wikipedia.com",
      "request": "request",
      "signatureCount": 0,
      "signatures": Array [],
      "status": "OPEN",
      "summary": "petition summary",
      "title": "Petition with summary, references, profile",
      "updatedAt": 2022-01-01T00:00:00.000Z,
    },
  ],
  "totalItems": 3,
}
`;

exports[`PostService listPosts returns posts ordered by signature count (least) 1`] = `
Object {
  "posts": Array [
    Object {
      "addressee": Object {
        "name": "Ministry of Environment",
        "shortName": "MoE",
      },
      "addresseeId": 1,
      "createdAt": 2022-01-01T00:00:00.000Z,
      "email": "John@email.com",
      "fullname": "John Doe",
      "hashedUserSgid": "wxyz",
      "id": "aaaaaaaa-dea2-4f78-8dc4-5542cafaff87",
      "profile": "OGP",
      "reason": "reason",
      "references": "wikipedia.com",
      "request": "request",
      "signatureCount": 0,
      "signatures": Array [],
      "status": "OPEN",
      "summary": "petition summary",
      "title": "Petition with summary, references, profile",
      "updatedAt": 2022-01-01T00:00:00.000Z,
    },
    Object {
      "addressee": Object {
        "name": "Ministry of Environment",
        "shortName": "MoE",
      },
      "addresseeId": 1,
      "createdAt": 2022-02-02T00:00:00.000Z,
      "email": "John@email.com",
      "fullname": "John Doe",
      "hashedUserSgid": "wxyz",
      "id": "395ac6ec-dea2-4f78-8dc4-1234cafaff87",
      "profile": null,
      "reason": "reason for",
      "references": null,
      "request": "request",
      "signatureCount": 0,
      "signatures": Array [],
      "status": "OPEN",
      "summary": null,
      "title": "Basic petition without optional items",
      "updatedAt": 2022-02-02T00:00:00.000Z,
    },
    Object {
      "addressee": Object {
        "name": "Ministry of Environment",
        "shortName": "MoE",
      },
      "addresseeId": 1,
      "createdAt": 2022-01-01T00:00:00.000Z,
      "email": "John@email.com",
      "fullname": "John Doe",
      "hashedUserSgid": "$2b$10$72TKINiR4.ZMaJZmuN5tW.uCuGZcA0TzLUo4qly55GyDndAYG8BSW",
      "id": "af7b6aa9-7c6e-4ea5-b536-4e55bc506847",
      "profile": null,
      "reason": "reason for",
      "references": null,
      "request": "request",
      "signatureCount": 3,
      "signatures": Array [
        Object {
          "comment": null,
          "createdAt": 2022-01-01T00:00:00.000Z,
          "fullname": null,
          "hashedUserSgid": "1234",
          "id": 1,
          "postId": "af7b6aa9-7c6e-4ea5-b536-4e55bc506847",
          "type": "support",
          "updatedAt": 2022-01-01T00:00:00.000Z,
        },
        Object {
          "comment": null,
          "createdAt": 2022-01-01T00:00:00.000Z,
          "fullname": "John Doe",
          "hashedUserSgid": "5678",
          "id": 2,
          "postId": "af7b6aa9-7c6e-4ea5-b536-4e55bc506847",
          "type": "support",
          "updatedAt": 2022-01-01T00:00:00.000Z,
        },
        Object {
          "comment": "I support this cause!",
          "createdAt": 2022-01-01T00:00:00.000Z,
          "fullname": null,
          "hashedUserSgid": "8765",
          "id": 3,
          "postId": "af7b6aa9-7c6e-4ea5-b536-4e55bc506847",
          "type": "support",
          "updatedAt": 2022-01-01T00:00:00.000Z,
        },
      ],
      "status": "OPEN",
      "summary": null,
      "title": "Basic petition without optional items",
      "updatedAt": 2022-01-01T00:00:00.000Z,
    },
  ],
  "totalItems": 3,
}
`;

exports[`PostService listPosts returns posts ordered by signature count 1`] = `
Object {
  "posts": Array [
    Object {
      "addressee": Object {
        "name": "Ministry of Environment",
        "shortName": "MoE",
      },
      "addresseeId": 1,
      "createdAt": 2022-01-01T00:00:00.000Z,
      "email": "John@email.com",
      "fullname": "John Doe",
      "hashedUserSgid": "$2b$10$72TKINiR4.ZMaJZmuN5tW.uCuGZcA0TzLUo4qly55GyDndAYG8BSW",
      "id": "af7b6aa9-7c6e-4ea5-b536-4e55bc506847",
      "profile": null,
      "reason": "reason for",
      "references": null,
      "request": "request",
      "signatureCount": 3,
      "signatures": Array [
        Object {
          "comment": null,
          "createdAt": 2022-01-01T00:00:00.000Z,
          "fullname": null,
          "hashedUserSgid": "1234",
          "id": 1,
          "postId": "af7b6aa9-7c6e-4ea5-b536-4e55bc506847",
          "type": "support",
          "updatedAt": 2022-01-01T00:00:00.000Z,
        },
        Object {
          "comment": null,
          "createdAt": 2022-01-01T00:00:00.000Z,
          "fullname": "John Doe",
          "hashedUserSgid": "5678",
          "id": 2,
          "postId": "af7b6aa9-7c6e-4ea5-b536-4e55bc506847",
          "type": "support",
          "updatedAt": 2022-01-01T00:00:00.000Z,
        },
        Object {
          "comment": "I support this cause!",
          "createdAt": 2022-01-01T00:00:00.000Z,
          "fullname": null,
          "hashedUserSgid": "8765",
          "id": 3,
          "postId": "af7b6aa9-7c6e-4ea5-b536-4e55bc506847",
          "type": "support",
          "updatedAt": 2022-01-01T00:00:00.000Z,
        },
      ],
      "status": "OPEN",
      "summary": null,
      "title": "Basic petition without optional items",
      "updatedAt": 2022-01-01T00:00:00.000Z,
    },
    Object {
      "addressee": Object {
        "name": "Ministry of Environment",
        "shortName": "MoE",
      },
      "addresseeId": 1,
      "createdAt": 2022-01-01T00:00:00.000Z,
      "email": "John@email.com",
      "fullname": "John Doe",
      "hashedUserSgid": "wxyz",
      "id": "aaaaaaaa-dea2-4f78-8dc4-5542cafaff87",
      "profile": "OGP",
      "reason": "reason",
      "references": "wikipedia.com",
      "request": "request",
      "signatureCount": 0,
      "signatures": Array [],
      "status": "OPEN",
      "summary": "petition summary",
      "title": "Petition with summary, references, profile",
      "updatedAt": 2022-01-01T00:00:00.000Z,
    },
    Object {
      "addressee": Object {
        "name": "Ministry of Environment",
        "shortName": "MoE",
      },
      "addresseeId": 1,
      "createdAt": 2022-02-02T00:00:00.000Z,
      "email": "John@email.com",
      "fullname": "John Doe",
      "hashedUserSgid": "wxyz",
      "id": "395ac6ec-dea2-4f78-8dc4-1234cafaff87",
      "profile": null,
      "reason": "reason for",
      "references": null,
      "request": "request",
      "signatureCount": 0,
      "signatures": Array [],
      "status": "OPEN",
      "summary": null,
      "title": "Basic petition without optional items",
      "updatedAt": 2022-02-02T00:00:00.000Z,
    },
  ],
  "totalItems": 3,
}
`;

exports[`PostService publishPost returns void on making post public 1`] = `
Object {
  "addresseeId": 1,
  "auditTrail": Object {},
  "createdAt": Any<Date>,
  "email": "John@email.com",
  "fullname": "John Doe",
  "hashedUserSgid": "wxyz",
  "id": "395ac6ec-dea2-4f78-8dc4-1234cafaff87",
  "profile": null,
  "reason": "reason for",
  "references": null,
  "request": "request",
  "salt": "$2b$10$72TKINiR4.ZMaJZmuN5tW.",
  "signatureOptions": Array [
    "support",
    "oppose",
  ],
  "status": "OPEN",
  "summary": null,
  "title": "Basic petition without optional items",
  "updatedAt": Any<Date>,
}
`;

exports[`PostService publishPost throws an error on invalid id 1`] = `"Post update failed"`;

exports[`PostService updatePost throws on bad addressee 1`] = `"Addressee does not exist"`;

exports[`PostService updatePost updates post on good input 1`] = `
Object {
  "addresseeId": 2,
  "auditTrail": Object {},
  "createdAt": Any<Date>,
  "email": "updated@email.com",
  "fullname": "John Doe",
  "hashedUserSgid": "$2b$10$72TKINiR4.ZMaJZmuN5tW.uCuGZcA0TzLUo4qly55GyDndAYG8BSW",
  "id": "af7b6aa9-7c6e-4ea5-b536-4e55bc506847",
  "profile": "User profile",
  "reason": "More compelling reason",
  "references": null,
  "request": "More compelling request",
  "salt": "$2b$10$72TKINiR4.ZMaJZmuN5tW.",
  "signatureOptions": Array [
    "support",
    "oppose",
  ],
  "status": "OPEN",
  "summary": "A longer description or summary",
  "title": "Title should be even more concise",
  "updatedAt": Any<Date>,
}
`;
